README.md
<img  align="center"  src="static/images/icons/icon-72x72.png"/> SPIDERWEB
===

### Ham radio cluster web viewer for DxSpider
[![GPLv3 license](https://img.shields.io/badge/License-GPLv3-blue.svg)](http://perso.crans.org/besson/LICENSE.html)
[![Linux](https://svgshare.com/i/Zhy.svg)](https://svgshare.com/i/Zhy.svg)
[![made-with-python](https://img.shields.io/badge/B.e.-Python-1f425f.svg)](https://www.python.org/)
[![made-with-javascript](https://img.shields.io/badge/F.e.-JavaScript-yellow)](https://www.javascript.com)
[![bootstrap](https://img.shields.io/badge/CSS%20Fwk-Bootstrap-purple)](https://getbootstrap.com/)
[![CodeFactor](https://www.codefactor.io/repository/github/coulisse/spiderweb/badge)](https://www.codefactor.io/repository/github/coulisse/spiderweb)


- **Release:** 
- **Author:** Corrado Gerbaldo - [IU1BOW](https://www.qrz.com/db/IU1BOW)
- **Mail:** <corrado.gerbaldo@gmail.com>
- **Licensing:** Gpl V3.0 see [LICENSE](LICENSE) file.
- **Languages:** This application is written in Python 3.11/flask,Javascript and HTML

___
**DXSpider** is a great DX Cluster software that has useful telnet interface. 
I wrote this application in order to add a web user interface to DXSpider and show the spots collected.
The user could see 50 spots at time and filter them by band, spotter continent and spotted continent.

For this application, I've used:
- **Bootstrap** for stylesheet CSS
- **Apache ECharts** for managing charts
- **qrz.com** For each callsing, found you can click on lens and you'll see him on qrz.com
- **flag-icon-css** [https://github.com/lipis/flag-icon-css](https://github.com/lipis/flag-icon-css) I used it to show the country flags 
- **ng3k.com** [ng3k.com](http://ng3k.com/misc/adxo.html) I used to get information about "Announced Dx Operations". Thanks to Bill/NG3K !!!
- **silso** [sidc.be/silso](https://sidc.be/silso/) used to show propagation trend in "Chart & stats" section

You can find my web site at [https://www.iu1bow.it](https://www.i1bow.it)

### Changelog
see file ["CHANGELOG.md"](docs/CHANGELOG.md)

### Install            

**1) DXSpider**
First of all you have to installed [DXspider](http://www.dxcluster.org/) and connected with some other cluster nodes.

**2) MariaDB / MySQL**
Then you have to install MariaDB on your web server, on the same server where DXSpider is running and configure DXSpider to use it: in your spider folder edit `local/DXVars.pm` adding these lines:
```DXWars.pm
# the SQL database DBI dsn
$dsn = "dbi:mysql:dxcluster:localhost:3306";
$dbuser = "your-user";
$dbpass = "your-password"; 
```
If you would change some MariaDB parameters, then you can find them in  `/etc/mysql/my.cnf` or `/etc/my.cnf`, depending on your distro.
If the database will not be created automatically, please see ["DB_ISSUES.md"](docs/DB_ISSUES.md) 

**3) Python modules**  
You could install python modules using automatic or manual way.

*3.1) Automatic modules install*  
after downloaded spiderweb move in the main folder and install using requirements.txt file
```console
 foo@bar:~$ cd spiderweb                                
 foo@bar:~$ pip install -r requirements.txt
```
*3.2) Manual modules install*  
First of all you have to install the python3 pip installer
```console
foo@bar:~$ sudo apt install python3-pip
```
This application is based on Flask 
To install **Flask**:
```console
foo@bar:~$ pip install flask 
foo@bar:~$ pip install Flask-minify
foo@bar:~$ pip install flask_wtf
foo@bar:~$ pip install pandas
```
Then you have to install mysql libraries**:
```console
foo@bar:~$ pip install mysql-connector-python
foo@bar:~$ pip install --upgrade mysql-connector-python==8.0.12

```

### Configuration  

In the path `spiderweb/cfg/` rename `config.json.template` in `config.json`:
```console
foo@bar:~$ mv config.json.template config.json
```
then edit it and set the user and password of your database, the menu items, and other stuffs (callsign, mail address...).
There is also a specific parameter, named "enable_cq_filter" used to enable the CQ Zone filtering.

Othewhise, if you prefer, you could use a utility to edit your configuration and menu. Go in "script" folder and run ./config.sh

```console
foo@bar:~$ cd scripts
foo@bar:~$ ./config.sh

*** DxSpider configuration ***
Configuration file loaded from: ../cfg/config.json

   h:  help
   vc: view config.
   ec: edit config.
   vm: view menu
   em: edit menu
   s:  save
   t:  load config. from template

   x:  exit

Make your choice: 

```

### Crontab
Starting from version 2.4, since all activities are managed by the application, you don't need to schedule anything.


### Run test
Now, you can run your web application with the following command:
```console
foo@bar:~$ python3 webapp.py
```
The flask default port is 5000, so you can see your web app, typing `http://localhost:5000` in your web browser.
Keep in mind that the flask web server, usually is used as a test server.

### Production
There are some ways to use it in production. 

My configuration is:
Cloudflare + Nginx + Bjoern

**Bjoern is a lightweight WSGI for python.**

for installing it: 
```console
foo@bar:~$ sudo apt install libev-dev libevdev2
foo@bar:~$ pip3 install bjoern
```

If you want you can make it as a _daemon service_.
Create and edit a file named for example spiderweb.service (in the systemd folder)

```console
foo@bar:~$ sudo nano /etc/systemd/system/spiderweb.service
```
Below an example of configuration, keep in mind that it is just an example (you have to put your data for your application [here a little guide](https://linuxhandbook.com/create-systemd-services/)   ):
```console
[Unit]
Description=bjoern instance spiderweb
After=network.target
After=multi-user.target

[Service]
User=web
Group=www-data
Type=simple
WorkingDirectory=/home/web/spiderweb
Environment="PATH=/home/web/spiderweb"
ExecStart=/usr/bin/python3 /home/web/spiderweb/wsgi.py

[Install]
WantedBy=multi-user.target
```

Then you can install and start the daemon:
```console
foo@bar:~$ sudo systemctl enable spiderweb.service
foo@bar:~$ sudo systemctl start spiderweb.service
foo@bar:~$ sudo systemctl status spiderweb.service

● spiderweb.service - bjoern instance spiderweb
   Loaded: loaded (/etc/systemd/system/spiderweb.service; enabled; vendor preset: enabled)
   Active: active (running) since Sun 2020-10-25 09:56:35 UTC; 8h ago
 Main PID: 6518 (python3)
    Tasks: 1 (limit: 420)
   CGroup: /system.slice/spiderweb.service
           └─6518 /usr/bin/python3 /home/web/spiderweb/wsgi.py

Oct 25 09:56:35 dxcluster01 systemd[1]: Started bjoern instance spiderweb.
```
Now you can install and configure **NGINX**

Install with:  
```console
foo@bar:~$ sudo apt install nginx
```

Configure:
```console
sudo nano /etc/nginx/sites-available/myapp
```

```console
server {
    listen 80;
    server_name iu1bow.it www.iu1bow.it;
    location ^~ /.well-known/ {
      alias /home/web/verify/.well-known/;
    }

    location / {
        ssi off;
        include proxy_params;
        proxy_pass http://localhost:8080/;
        proxy_set_header Host $host;
    }
}

```

For SSL, I'm using [Cloudflare](https://www.cloudflare.com/). This is a free service that allow you to use https and a proxy cache. 


**Search engine indexing:** when you are on-line, if you would to index your website on search engines, you have to generate a file named sitemap.xml and put it in /static/ folder. There are many tools to generate sitemap.xml, for example [www.xml-sitemaps.com](https://www.xml-sitemaps.com/)

**Index on MySQL:** if you would to increase speed on callsign search, you could define some index on the table 'spot'. You can see more details on ["create_mysql_index.sql"](scripts/create_mysql_index.sql)


### Mobile
This application is designed for desktop and mobile phone. It is a [PWA](https://en.wikipedia.org/wiki/Progressive_web_app), so it could be installed and used like an app on mobile.

### API
**Spot list**

You can retrieve last spots calling "**/spotlist**"; For example  [www.iu1bow.com/spotlist](https://www.iu1bow.com/spotlist)

**Country of a callsign**

You can retrive some information about a callsign with **callsign**; for example:  [www.iu1bow.com/callsign?c=IU1BOW](https://www.iu1bow.com/callsign?c=IU1BOW)


### Development
**Directory structure**
```
/                 . main application files
├── cfg           . configuration files (put here your config.json with your setting)
├── docs          . documentation
├── lib           . python libs used for the application
├── log           . application log
├── scripts       . utility scripts for testing, build etc.
├── static        . static files css, js, data, html, images etc.
│   ├── css       .
│   │   ├── dev   . development css not minifyed/uglifyed
│   │   └── rel   . release css minifyed/uglifyed  (do not change these files)
│   ├── data      . application data (world.json)
│   ├── html      .
│   │   └── dev   . html templates used for build release static html (for offline)
│   │   └── rel   . release static html (for offline)
│   ├── images    . static images
│   │   └── icons . static icons
│   └── js        .
│       ├── dev   . development js not minifyed/uglifyed
│       └── rel   . release js minifyed/uglifyed  (do not change these files)
└── templates     . html templates used by python flask for serve the web pages
```
**Application description**

The main **server** application ```webapp.py``` is in the root folder. In this application there are routing to html dynamic templates and serves also back-end API. This is wrapped by ```wsgi.py``` for using with **bjoern** server.

Static files (css, js...) are in ```static``` directory: here there are subdirectories: 
- ```dev```  where you can edit and modify sources
- ```rel```  here there are release files created with the building process and used in production

**Lint**

For lint javascript I use **ESLint**. You can install with ```npm init @eslint/config```
 pylint        ```pip install pylint```

**Building process**

Prerequisites:
| **Component** | **Description**                                    | **Install command**             |
|---------------|----------------------------------------------------|---------------------------------|

docs/CHANGELOG.md
### Change log
Date: 12/09/2023 
Release: 
- changed cache-control header
- fixed Layout scrolling (SEO)
- first time spot load: not show cyan background
- efactor of the plot code
- fixed issue #31
- added PWA cache and improved service worker for offline

___
Date: 25/01/2023 
Release: v2.4.1.2
- fixed issue #30  bug on callsign search

___
Date: 23/01/2023 
Release: v2.4.1.1
- fixed issue #29  bug on callsign search

___
Date: 19/01/2023 
Release: v2.4.1
- changed dimensions of spots in world dx spost charts
- managed empty data in data providers for charts
- removed jQuery: migrated to vanilla javascript
- for spot refresh get only new spots starting from last rowid
- modified building script 
- moved cty file in data directory
- fixed issue #28
- cut comments too long
- charts layout changed
- reduced javascripts
- fixed issue #11 cookie consent request for non-https website

___
Date: 01/01/2023 
Release: v2.4
- migration to python 3.11
- added descriptions to continents
- fixed issue #23: Wrong flag for Cocos (Keeling) Islands
- proposal for improvements #25 (charts titles, removed link from connected nodes)
- reenginering of graph & stats
- upgraded bootstrap icons
- upgraded flag icons
- added new propagation chart https://sidc.be/silso/
- added Content Security Policy
___
Date: 28/09/2022 
Release: v2.3.4
- fixed issue #22 propagation_heatmaps.sh fails with 'Passing a Normalize instance simultaneously with vmin/vmax is not supported.' 
- replaced seaborn styles since are deprecated
___
Date: 23/08/2022                         
Release: v2.3.3
- modified minified system
- fixed qso_world_plot.py that hangs
___
Date: 23/08/2022                         
Release: v2.3.2
- upgraded Numpy library for security issues
- upgraded Mysql library for security issues
- upgraded to bootstrap 5.2
- upgraded flag-icon-css to 6.6.4
___
Date: 28/04/2022                         
Release: v2.3.1
- Fixed bug on adxo json that block entire website!
- upgraded Pillow library for security issues
___
Date: 04/03/2022                         
Release: v2.3
- lint on code with codefactor
- added new world qso in lasth month plot
___
Date: 13/01/2022                         
Release: v2.2
- fixed country code of Curacao
- managed CTY.dat file for a more precise localization of callsign
- bootstrap upgraded to 5.1.3
- added UTC clock in menu bar
- added cookie consent banner
- updated flag-icon-css
- modified modes.json in order to improve modes filter
- fixed forecast on  months plots
- upgraded pillow library to 9.0 for security issues
___
Date: 04/12/2021                         
Release: v2.1
- created a new "spider" icon
- jquery upgraded to 3.6
- bootstrap upgraded to 5.0.2
- managed connection error to telnet host 
- used a base template in order to put all csv/scripts on a page
- keywords changed
- added version to page footer
- created a build script in order to automating versioning and others
- added security policy in meta tag
- added cross-site request forgery security
- added security headers
- added cache control headers
___
16/05/2021: 2.0.2   
- Fixed frequency mode on 40 meters                                                    
- Fixed unicode query
- used qry.py also inside webapp.py
- replaced MySqlDb with mysql.connector
___
14/05/2021: 2.0.1   
- Addded installation with requirements                                                
___
06/05/2021: 2.0     
- Migrated to bootstrap 5                                                               
- Migrated to python 3
- Added configuration in config.json for telnet and mail address (so you can avoid to change index.html file)
- In the spot list, moved the country columns, after flag
- When press "search", collapse filters form
- Added mode filter
- fixed label on qso/months plot
- removed link to country-flags.io and used https://github.com/lipis/flag-icon-css hosted at https://cdnjs.com/libraries/flag-icon-css/3.4.1 in order to improve performances and remove cookies
- removed cookie bar
- improved offline page
- remove colors from callsing search resoults table
- added popover on mobile
- added monitor.sh in order to monitor your system
- added connected nodes in plots page (renamed as "plots and stats")
- added hour / band qso plot
- improved qso per month plot
- added plot trend
- improved logging
- insert bootstrap icons
- improved footer
- added "Announced Dx Operation" integration with ng3k website 
- used bootstrap icons instead of custom icons
- migrated to last jquery version 3.6.0
___
04/10/2020: 1.2.2
- Added config.sh utilty in script folder, in order to avoid edit manually config.json
- fixex menu on callsign page
___
21/09/2020: 1.2.1
- Added country column (only on desktop version)
- Added external configurable menu
- Renamed json.config.example json.config.template
- Added code of conduct and issue templates
___
08/09/2020: 1.2    
- Modified cookies.html in order to insert dinamically the host name (instead of dxcluster.iu1bobw.it)
- Removed sitemap.xml from git
- updated jquery version to 3.5.1
- added callsign search
- made flags responsive
- some little changes on meta tag and pages descriptions

webapp.py
__author__ = "IU1BOW - Corrado"
import flask
import secrets
from flask import request, render_template
from flask_wtf.csrf import CSRFProtect
from flask_minify import minify
import json
import threading
import logging
import logging.config
from lib.dxtelnet import who
from lib.adxo import get_adxo_events
from lib.qry import query_manager
from lib.cty import prefix_table
from lib.plot_data_provider import ContinentsBandsProvider, SpotsPerMounthProvider, SpotsTrend, HourBand, WorldDxSpotsLive
import calendar
import time


logging.config.fileConfig("cfg/webapp_log_config.ini", disable_existing_loggers=True)
logger = logging.getLogger(__name__)
logger.info("Start")

app = flask.Flask(__name__)
app.config["SECRET_KEY"] = secrets.token_hex(16)
app.config.update(
    SESSION_COOKIE_SECURE=True,
    SESSION_COOKIE_HTTPONLY=False,
    SESSION_COOKIE_SAMESITE="Strict",
)

inline_script_nonce = ""

csrf = CSRFProtect(app)
logger.debug(app.config)

if app.config["DEBUG"]:
    minify(app=app, html=False, js=False, cssless=False)
else:
    minify(app=app, html=True, js=True, cssless=False)

# load config file
with open("cfg/config.json") as json_data_file:
    cfg = json.load(json_data_file)

logging.debug("CFG:")
logging.debug(cfg)
# load bands file
with open("cfg/bands.json") as json_bands:
    band_frequencies = json.load(json_bands)

# load mode file
with open("cfg/modes.json") as json_modes:
    modes_frequencies = json.load(json_modes)

# load continents-cq file
with open("cfg/continents.json") as json_continents:
    continents_cq = json.load(json_continents)
    
# load heatmap file
with open("cfg/heatmap.json") as json_heatmap:
    heatmap_data = json.load(json_heatmap)

# read and set default for enabling cq filter
if cfg.get("enable_cq_filter"):
    enable_cq_filter = cfg["enable_cq_filter"].upper()
else:
    enable_cq_filter = "N"

# define country table for search info on callsigns
pfxt = prefix_table()

# create object query manager
qm = query_manager()

# find id  in json : ie frequency / continent
def find_id_json(json_object, name):
    return [obj for obj in json_object if obj["id"] == name][0]


def query_build_callsign(callsign):

    # query_string = ""
    # if len(callsign) <= 14:
    last_rowid = request.args.get("lr")
    if last_rowid is None:
            last_rowid = "0"
    if not last_rowid.isnumeric():
        last_rowid = 0
    items = callsign.split(",")
    search_string = "','".join(items)
    search_string = "'" + search_string + "'"
    # query_string = (
    #     "(SELECT rowid, spotter AS de, freq, spotcall AS dx, comment AS comm, time, spotdxcc from spot WHERE spotter=("
    #     + search_string
    #     + ""
    # )
    # query_string += " ORDER BY rowid desc limit 10)"
    # query_string += " UNION "
    query_string = (
        "(SELECT rowid, spotter AS de, freq, spotcall AS dx, comment AS comm, time, spotdxcc from spot WHERE rowid > " + last_rowid + " && spotcall IN ("
        + search_string
        + ")"
    )
    query_string += " ORDER BY rowid desc limit 50);"
    # else:
    #     logging.warning("callsign too long")
    return query_string


def query_build():

    try:
        # get url parameters
        last_rowid = request.args.get("lr")  # Last rowid fetched by front end
        callsign = request.args.get("c")  # search specific callsign
        band = request.args.getlist("b")  # band filter
        dere = request.args.getlist("e")  # DE continent filter
        dxre = request.args.getlist("x")  # Dx continent filter
        mode = request.args.getlist("m")  # mode filter
        exclft8 = request.args.getlist("exclft8")  # Mode DIGI explode FT8
        exclft8 = exclft8[0]
        exclft4 = request.args.getlist("exclft4")  # Mode DIGI explode FT4
        exclft4 = exclft4[0]
        decq = request.args.getlist("qe")  # DE cq zone filter
        dxcq = request.args.getlist("qx")  # DX cq zone filter

        query_string = ""

        items = callsign.split(",")
        search_string = "','".join(items)
        search_string = "'" + search_string + "'"
        callsign_qry_string = " AND (spotcall IN (" + search_string + "))"

        # construct band query decoding frequencies with json file
        band_qry_string = " AND (("
        for i, item_band in enumerate(band):
            freq = find_id_json(band_frequencies["bands"], item_band)
            if i > 0:
                band_qry_string += ") OR ("

            band_qry_string += (
                "freq BETWEEN " + str(freq["min"]) + " AND " + str(freq["max"])
            )

        band_qry_string += "))"

        # construct mode query
        mode_qry_string = " AND  (("
        for i, item_mode in enumerate(mode):
            single_mode = find_id_json(modes_frequencies["modes"], item_mode)
            if i > 0:
                mode_qry_string += ") OR ("
            for j in range(len(single_mode["freq"])):
                if j > 0:
                    mode_qry_string += ") OR ("
                mode_qry_string += (
                    "freq BETWEEN "
                    + str(single_mode["freq"][j]["min"])
                    + " AND "
                    + str(single_mode["freq"][j]["max"])
                )
        mode_qry_string += "))"       

        ft8_qry_string = " AND ("
        if exclft8 == "true":
            ft8_qry_string += "(comment NOT LIKE '%FT8%')"
            single_mode = find_id_json(modes_frequencies["modes"], "digi-ft8")
            for j in range(len(single_mode["freq"])):
                ft8_qry_string += (
                    " AND (freq NOT BETWEEN "
                    + str(single_mode["freq"][j]["min"])
                    + " AND "
                    + str(single_mode["freq"][j]["max"])
                    + ")"
                )
        ft8_qry_string += ")" 

        ft4_qry_string = " AND ("
        if exclft4 == "true":
            ft4_qry_string += "(comment NOT LIKE '%FT4%')"
            single_mode = find_id_json(modes_frequencies["modes"], "digi-ft4")
            for j in range(len(single_mode["freq"])):
                ft4_qry_string += (
                    " AND (freq NOT BETWEEN "
                    + str(single_mode["freq"][j]["min"])
                    + " AND "
                    + str(single_mode["freq"][j]["max"])
                    + ")"
                )
        ft4_qry_string += ")" 

        # construct DE continent region query
        dere_qry_string = " AND spottercq IN ("
        for i, item_dere in enumerate(dere):
            continent = find_id_json(continents_cq["continents"], item_dere)
            if i > 0:
                dere_qry_string += ","
            dere_qry_string += str(continent["cq"])
        dere_qry_string += ")"

        # construct DX continent region query
        dxre_qry_string = " AND spotcq IN ("
        for i, item_dxre in enumerate(dxre):
            continent = find_id_json(continents_cq["continents"], item_dxre)
            if i > 0:
                dxre_qry_string += ","
            dxre_qry_string += str(continent["cq"])
        dxre_qry_string += ")"

        if enable_cq_filter == "Y":
            # construct de cq query
            decq_qry_string = ""
            if len(decq) == 1:
                if decq[0].isnumeric():
                    decq_qry_string = " AND spottercq =" + decq[0]
            # construct dx cq query
            dxcq_qry_string = ""
            if len(dxcq) == 1:
                if dxcq[0].isnumeric():
                    dxcq_qry_string = " AND spotcq =" + dxcq[0]

        if last_rowid is None:
            last_rowid = "0"
        if not last_rowid.isnumeric():
            last_rowid = 0

        query_string = (
            "SELECT rowid, spotter AS de, freq, spotcall AS dx, comment AS comm, time, spotdxcc from spot WHERE rowid > "
            + last_rowid
        )

        if callsign:
            query_string += callsign_qry_string
            
        if len(band) > 0:
            query_string += band_qry_string

        if len(mode) > 0:
            query_string += mode_qry_string

        if exclft8 == "true":
            query_string += ft8_qry_string

        if exclft4 == "true":
            query_string += ft4_qry_string

        if len(dere) > 0:
            query_string += dere_qry_string

        if len(dxre) > 0:
            query_string += dxre_qry_string

        if enable_cq_filter == "Y":
            if len(decq_qry_string) > 0:
                query_string += decq_qry_string

            if len(dxcq_qry_string) > 0:
                query_string += dxcq_qry_string

        query_string += " ORDER BY rowid desc limit 50;"

    except Exception as e:
        logger.error(e)
        query_string = ""
        
    return query_string


# the main query to show spots
# it gets url parameter in order to apply the build the right query
# and apply the filter required. It returns a json with the spots
def spotquery():
    try:

        # callsign = request.args.get("c")  # search specific callsign

        # if callsign:
        #     query_string = query_build_callsign(callsign)
        # else:
        #     query_string = query_build()
        query_string = query_build()
        qm.qry(query_string)
        data = qm.get_data()
        
        row_headers = qm.get_headers()

        logger.debug("query done")
        logger.debug(data)

        if data is None or len(data) == 0:
            logger.warning("no data found")

        payload = []
        for result in data:
            # create dictionary from recorset
            main_result = dict(zip(row_headers, result))
            # find the country in prefix table
            search_prefix = pfxt.find(main_result["dx"])
            # merge recordset and contry prefix
            main_result["country"] = search_prefix["country"]
            main_result["iso"] = search_prefix["iso"]

            payload.append({**main_result})

        return payload
    except Exception as e:
        logger.error(e)

def get_dx_calls():
    try:
        query_string = "SELECT * FROM (SELECT spotcall AS dx FROM spot ORDER BY rowid DESC LIMIT 0, 500) sub1 GROUP BY dx"
        qm.qry(query_string)
        data = qm.get_data()
        row_headers = qm.get_headers()

        payload = []
        for result in data:
            main_result = dict(zip(row_headers, result))
            payload.append(main_result["dx"])

        return payload
    except Exception as e:
        return []
    
def heatmapquery():
    try:

        de = request.args.getlist("c")  # DE continent filter
        de = de[0]
        
        query_string = "SELECT continent, band, COUNT(*) AS `count` FROM ("
        query_string += "SELECT spotter, continent, band FROM ("
        query_string += "SELECT spotter, "

        query_string += "CASE "
        for obj in continents_cq["continents"]:
            query_string += "WHEN spotcq IN (" + obj["cq"] + ")" + " THEN '" + obj["id"] + "' "    
        query_string += "END AS continent, "

        query_string += "CASE "
        for obj in heatmap_data["bands"]:
            query_string += "WHEN freq BETWEEN " + str(obj["min"]) + " AND " + str(obj["max"]) + " THEN '" + obj["id"] + "' "    
        query_string += "END AS band "

        query_string += "FROM spot WHERE "
        
        current_GMT = time.gmtime()
        time_stamp = calendar.timegm(current_GMT)
        time_stamp = time_stamp - 60 * 60
        query_string += "time >= " + str(time_stamp) + " && "

        query_string += "spottercq IN ("
        continent = find_id_json(continents_cq["continents"], de)
        query_string += str(continent["cq"])
        query_string += ")) sub1 WHERE continent IS NOT NULL && band IS NOT NULL GROUP BY spotter) sub2 GROUP BY continent, band"

        qm.qry(query_string)
        data = qm.get_data()
        row_headers = qm.get_headers()

        payload = []
        maxValue = 0
        for result in data:
            row = dict(zip(row_headers, result))
            payload.append({
                "x": heatmap_data["x"][row["band"]],
                "y": heatmap_data["y"][row["continent"]],
                "value": row["count"]
            })
            if maxValue < row["count"]:
                maxValue = row["count"]
        payload.append({
            "max": maxValue
        })

        return payload
    except Exception as e:
        logger.error(e)

# find adxo events
adxo_events = None


def get_adxo():
    global adxo_events
    adxo_events = get_adxo_events()
    threading.Timer(12 * 3600, get_adxo).start()


get_adxo()

# create data provider for charts
heatmap_cbp = ContinentsBandsProvider(logger, qm, continents_cq, band_frequencies)
bar_graph_spm = SpotsPerMounthProvider(logger, qm)
line_graph_st = SpotsTrend(logger, qm)
bubble_graph_hb = HourBand(logger, qm, band_frequencies)
geo_graph_wdsl = WorldDxSpotsLive(logger, qm, pfxt)

# ROUTINGS
@app.route("/spotlist", methods=["GET"])
def spotlist():
    response = flask.Response(json.dumps(spotquery()))
    return response

@app.route("/heatmap", methods=["GET"])
def heatmap():
    response = flask.Response(json.dumps(heatmapquery()))
    return response


def who_is_connected():
    host_port = cfg["telnet"].split(":")
    response = who(host_port[0], host_port[1], cfg["mycallsign"])
    return response

#Calculate nonce token used in inline script and in csp "script-src" header
def get_nonce():
    global inline_script_nonce
    inline_script_nonce = secrets.token_hex()
    return inline_script_nonce

@app.route("/", methods=["GET"])
@app.route("/index.html", methods=["GET"])
def spots():
    response = flask.Response(
        render_template(
            "index.html",
            page='index',
            inline_script_nonce=get_nonce(),
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
            enable_cq_filter=enable_cq_filter,
            timer_interval=cfg["timer"]["interval"],
            adxo_events=adxo_events,
            continents=continents_cq,
            bands=band_frequencies,
            dx_calls=get_dx_calls(),
        )
    )
    return response


@app.route("/service-worker.js", methods=["GET"])
def sw():
    return app.send_static_file("pwa/service-worker.js")

@app.route("/offline.html")
def root():
    return app.send_static_file("html/offline.html")


@app.route("/world.json")
def world_data():
    return app.send_static_file("data/world.json")

@app.route("/propagation.html", methods=["GET"])
def propagation():
    response = flask.Response(
        render_template(
            "propagation.html",
            page='propagation',
            inline_script_nonce=get_nonce(),          
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
        )
    )
    return response

@app.route("/plots.html")
def plots():
    whoj = who_is_connected()
    response = flask.Response(
        render_template(
            "plots.html",
            page='plots',
            inline_script_nonce=get_nonce(),          
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
            who=whoj,
            continents=continents_cq,
            bands=band_frequencies,
        )
    )
    return response


@app.route("/cookies.html", methods=["GET"])
def cookies():
    response = flask.Response(
        render_template(
            "cookies.html",
            page='cookies',
            inline_script_nonce=get_nonce(),          
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
        )
    )
    return response


@app.route("/privacy.html", methods=["GET"])
def privacy():
    response = flask.Response(
        render_template(
            "privacy.html",
            page='privacy',
            inline_script_nonce=get_nonce(),          
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
        )
    )
    return response


@app.route("/sitemap.xml")
def sitemap():
    return app.send_static_file("sitemap.xml")


@app.route("/callsign.html", methods=["GET"])
def callsign():
    # payload=spotquery()
    callsign = request.args.get("c")
    response = flask.Response(
        render_template(
            "callsign.html",
            page='callsign',
            inline_script_nonce=get_nonce(),              
            mycallsign=cfg["mycallsign"],
            telnet=cfg["telnet"],
            mail=cfg["mail"],
            menu_list=cfg["menu"]["menu_list"],
            timer_interval=cfg["timer"]["interval"],
            callsign=callsign,
            adxo_events=adxo_events,
            continents=continents_cq,
            bands=band_frequencies,
        )
    )
    return response


# API that search a callsign and return all informations about that
@app.route("/callsign", methods=["GET"])
def find_callsign():
    callsign = request.args.get("c")
    response = pfxt.find(callsign)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.route("/plot_get_heatmap_data", methods=["GET"])
def get_heatmap_data():
    continent = request.args.get("continent")
    response = flask.Response(json.dumps(heatmap_cbp.get_data(continent)))
    logger.debug(response)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.route("/plot_get_dx_spots_per_month", methods=["GET"])
def get_dx_spots_per_month():
    response = flask.Response(json.dumps(bar_graph_spm.get_data()))
    logger.debug(response)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.route("/plot_get_dx_spots_trend", methods=["GET"])
def get_dx_spots_trend():
    response = flask.Response(json.dumps(line_graph_st.get_data()))
    logger.debug(response)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.route("/plot_get_hour_band", methods=["GET"])
def get_dx_hour_band():
    response = flask.Response(json.dumps(bubble_graph_hb.get_data()))
    logger.debug(response)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.route("/plot_get_world_dx_spots_live", methods=["GET"])
def get_world_dx_spots_live():
    response = flask.Response(json.dumps(geo_graph_wdsl.get_data()))
    logger.debug(response)
    if response is None:
        response = flask.Response(status=204)
    return response


@app.context_processor
def inject_template_scope():
    injections = dict()

    def cookies_check():
        value = request.cookies.get("cookie_consent")
        return value == "true"

    injections.update(cookies_check=cookies_check)
    return injections


@app.after_request
def add_security_headers(resp):

    resp.headers["Strict-Transport-Security"] = "max-age=1000"
    resp.headers["X-Xss-Protection"] = "1; mode=block"
    resp.headers["X-Frame-Options"] = "SAMEORIGIN"
    resp.headers["X-Content-Type-Options"] = "nosniff"
    resp.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
    resp.headers["Cache-Control"] = "public, no-cache"
    resp.headers["Pragma"] = "no-cache"

    
    
    resp.headers["Content-Security-Policy"] = "\
    default-src 'self';\
    script-src 'self' cdnjs.cloudflare.com cdn.jsdelivr.net 'nonce-"+inline_script_nonce+"' https://www.gstatic.com;\
    style-src 'self' cdnjs.cloudflare.com cdn.jsdelivr.net 'unsafe-inline';\
    object-src 'none';base-uri 'self';\
    connect-src 'self' cdn.jsdelivr.net cdnjs.cloudflare.com sidc.be;\
    font-src 'self' cdn.jsdelivr.net;\
    frame-src https://grafana.gafner.net https://muf.hb9vqq.ch;\
    frame-ancestors https://grafana.gafner.net https://muf.hb9vqq.ch;\
    form-action 'none';\
    img-src 'self' data: cdnjs.cloudflare.com sidc.be;\
    manifest-src 'self';\
    media-src 'self';\
    worker-src 'self';\
    "
    return resp
   
    #script-src 'self' cdnjs.cloudflare.com cdn.jsdelivr.net 'nonce-sedfGFG32xs';\
    #script-src 'self' cdnjs.cloudflare.com cdn.jsdelivr.net 'nonce-"+inline_script_nonce+"';\

if __name__ == "__main__":
    app.run(host="0.0.0.0")
